### PR TITLE
Add systray functionability

### PR DESCRIPTION
**Pull request by tpruzina**
As per discussion in #83 I have implemented simple tray icon functionality.

Currently there are 3 new UI checkbuttons+configs in Configuration tab (enable trayicon, start in tray, close/minimize to tray), those allow pavucontrol-qt to be run-able in background after boot.

As far as usage goes:
* (double or single) Lclick on systray toggles device (un)mute
* scrolling up/down over trayicon increases/decreases volume of all sinks
* middle mouse button shows/hides main window
* Rclick on trayicon shows simple context menu that  allows displaying/hiding as well as force exiting the application regardless of whether close to tray option is enabled

As far as audio implementation goes, I have tried to avoid copying pulse code, this uses already existing UI elements to modify the state ([[1]](https://github.com/lxqt/pavucontrol-qt/compare/master...tpruzina:systray-pr?expand=1#diff-28cf5512931be215bb86138f9f05e8cdR1254) ,[[2]](https://github.com/lxqt/pavucontrol-qt/compare/master...tpruzina:systray-pr?expand=1#diff-28cf5512931be215bb86138f9f05e8cdR1273)).

For this PR I squashed this into single commit, although whole commit chain is available [here](https://github.com/lxqt/pavucontrol-qt/compare/master...tpruzina:master).

I would also like to add single instance functionality (simplest way is probably qtsingleapplication but that introduces qtnetwork dependency iirc) and possibly commandline argument handling for systray, but hopefully this is already reviewable.

### CODE REVIEW

File: src/mainwindow.cc [hunk 1]
```diff
@@ -1170,3 +1200,139 @@ void MainWindow::onShowVolumeMetersCheckButtonToggled(bool toggled) {
...
+        double single_percent_step = (double)(max-min) / 100;
+        vs->setValue(vol + step*single_percent_step);
+    }
+}
+
+void MainWindow::iconActivated(QSystemTrayIcon::ActivationReason reason){
+    switch(reason)
+    {
+    case QSystemTrayIcon::Trigger:
+    case QSystemTrayIcon::DoubleClick:
```
Comment (palinek) [inline src/mainwindow.cc]: Just a matter of taste.... but I would suggest left-click -> show/hide window, middle-click-> mute/unmute.

File: src/mainwindow.cc [hunk 2]
```diff
@@ -119,6 +122,30 @@ MainWindow::MainWindow():
     if (sourceTypeSelection.isValid())
         sourceTypeComboBox->setCurrentIndex(sourceTypeSelection.toInt());
 
+    const QVariant useSystray = config.value("systray/enabled");
```
Comment (palinek) [inline src/mainwindow.cc]: Did you consider using a default value like in other places (e.g. `config.value("systray/enabled", false).toBool()`)?

**Comment by palinek** [comment]
> I would also like to add single instance functionality

Please, don't do that... -> https://github.com/lxqt/pavucontrol-qt/pull/17

File: src/mainwindow.cc (hunk 1 above)
Comment (tpruzina) [reply src/mainwindow.cc]: Makes sense.

File: src/mainwindow.cc (hunk 2 above)
Comment (tpruzina) [reply src/mainwindow.cc]: Will fix that for clarity, to be honest I didn't notice default value in the line above that.

File: src/mainwindow.cc [hunk 3]
```diff
@@ -119,6 +122,30 @@ MainWindow::MainWindow():
     if (sourceTypeSelection.isValid())
         sourceTypeComboBox->setCurrentIndex(sourceTypeSelection.toInt());
 
+    const QVariant useSystray = config.value("systray/enabled", false);
+    if(useSystray.isValid()) {
```
[the remark this answers is not in the document]
Comment (tpruzina) [reply src/mainwindow.cc]: Gotcha, used those just to maintain style with existing code, wasn't sure if there could be any garbage in boolean conversions.

**Comment by tpruzina** [comment]
@palinkek Assuming this looks OK, should I resubmit squashed version as single commit PR?

File: src/mainwindow.cc [hunk 4]
```diff
@@ -95,6 +96,8 @@ MainWindow::MainWindow():
     quit->setShortcut(QKeySequence::Quit);
     addAction(quit);
 
+    qApp->installEventFilter(this);
```
Comment (palinek) [inline src/mainwindow.cc]: Why installing the event filter on whole application? We just need the systray, not?

File: src/mainwindow.cc [hunk 5]
```diff
@@ -1170,3 +1187,136 @@ void MainWindow::onShowVolumeMetersCheckButtonToggled(bool toggled) {
...
+
+void MainWindow::createTrayIcon() {
+    if (!QSystemTrayIcon::isSystemTrayAvailable())
+        return;
+
+    systrayQuitAction.setText(tr("&Quit"));
+    systrayRestoreAction.setText(tr("&Restore"));
+    systrayMinimizeAction.setText(tr("Mi&nimize"));
+
+    connect(&systrayMinimizeAction, SIGNAL(triggered()), this, SLOT(hide()));
```
Comment (palinek) [inline src/mainwindow.cc]: Please, use the c++11 signal/slot notation

File: src/mainwindow.cc [hunk 6]
```diff
@@ -119,6 +122,17 @@ MainWindow::MainWindow():
...
+    enableSystrayCheckButton->setChecked(config.value("systray/enabled", false).toBool());
+    startInSystrayCheckButton->setChecked(config.value("systray/startInTray", false).toBool());
+    closeToSystrayCheckButton->setChecked(config.value("systray/closeToTray", false).toBool());
+    
+    if(!enableSystrayCheckButton->isChecked()) {
+        startInSystrayCheckButton->setDisabled(true);
+        closeToSystrayCheckButton->setDisabled(true);
+    }
+
+    createTrayIcon();
```
Comment (palinek) [inline src/mainwindow.cc]: Could the systray icon be created/deleted only when it is needed?

File: src/mainwindow.cc (hunk 4 above)
Comment (tpruzina) [reply src/mainwindow.cc]: Fixed this by decorating QSystemTrayIcon class (eventFilter() method necessary to make this work), not quite sure whether I should start moving more stuff into it or not (most Systray code equally codependent with MainWindow).

File: src/mainwindow.cc (hunk 6 above)
Comment (tpruzina) [reply src/mainwindow.cc]: This ties to previous, should I make [Systray](https://github.com/lxqt/pavucontrol-qt/pull/93/files#diff-e53fc39ebc111e4e739351226712e953R44) into a real class in separate CO and move everything there?

File: src/mainwindow.h [hunk 7]
```diff
@@ -100,11 +108,14 @@ protected Q_SLOTS:
...
     void deleteEventRoleWidget();
 
+    void setVisible(bool visibility);
+
     QByteArray defaultSinkName, defaultSourceName;
 
     bool canRenameDevices;
 
 private:
+    Systray *systray;
```
Comment (palinek) [inline src/mainwindow.h]: Please, use some kind of smart pointer

File: src/systray.cc [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    systrayIconMenu.addAction(&systrayQuitAction);
+
+    QSize sz(256,256);
+    systrayIcons[NOT_MUTED].addPixmap(mw.style()->standardIcon(QStyle::SP_MediaVolume).pixmap(sz));
+    systrayIcons[MUTED].addPixmap(mw.style()->standardIcon(QStyle::SP_MediaVolumeMuted).pixmap(sz));
+
+    setIcon(systrayIcons[NOT_MUTED]);
+    setContextMenu(&systrayIconMenu);
+    
+    qApp->installEventFilter(this);
```
Comment (palinek) [inline src/systray.cc]: Why are you installing event filter for all events ever... you just need/want to get your own events. So you just need to override `void event(QEvent *e)` and not to check every one event of the whole application. Not?

**Comment by tpruzina** [comment]
@palinek fixed latest 2 issues you brought up, sorry that it took some time, got busy with life.

**Comment by agaida** [comment]
???? - I'm only a bit irritated that there is no closing reason.

**Comment by tpruzina** [comment]
Closing this due to lack of interest and feedback.

**Pull request closed without merging**